Fix ValidatingAdmissionPolicy param resolution race condition

afshin-paydar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
It fixes a race condition in ValidatingAdmissionPolicy param resolution that causes false "no params found" errors when resources are created in batch

Problem:
When a ConfigMap, ValidatingAdmissionPolicy, and ValidatingAdmissionPolicyBinding are created in quick succession, the admission request may fail with:
```
  failed to configure binding: no params found for policy binding with `Deny` parameterNotFoundAction
```

This happens even though the ConfigMap exists, because:
  1. The ConfigMap informer was already synced (from earlier startup)
  2. `WaitForCacheSync()` returns immediately since the informer completed its initial LIST
  3. The watch event for the newly created ConfigMap hasn't propagated to the cache yet
  4. `paramStore.Get()` returns NotFound, causing the policy to deny the request

Solution:
Wire in `dynamicClient` and `restMapper` to enable direct API fallback when the informer cache returns NotFound:
  - **Fast path:** Try informer cache first (instant, no API server load)
  - **Fallback:** On NotFound error, make a single direct GET to the API server
  - **Self-correcting:** Once cache syncs, all future requests use the fast path

#### Which issue(s) this PR is related to:
https://github.com/kubernetes/kubernetes/issues/133827
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

  **Test coverage:**
  Added `TestCollectParamsHandlesCacheMiss` that:
  - Creates a ConfigMap directly in the API server (bypassing informer cache)
  - Verifies `CollectParams` successfully retrieves it via client fallback
  - Without this fix, the test would fail with NotFound error

  **Changes:**
  1. Added `DynamicClient` and `RESTMapper` fields to `PolicyHook` struct
  2. Modified `CollectParams` to accept these parameters
  3. Implemented `getParamDirectly` helper for direct API calls
  4. Updated call sites to pass the new parameters


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aaron-prindle commented:
/cc @cici37

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:
```diff
@@ -331,7 +331,38 @@ func CollectParams(
+		var param runtime.Object
+		var err error
+		const maxRetries = 3
+		retryDelay := 50 * time.Millisecond
```
I'm concerned that, over time, there will be additional changes requests to increase this limit to better handle control plane configurations where the the cache is farther behind.

How about, instead, we wire in a client and make a direct get as a fallback if we have a cache miss?

jpbetz reviewed (commented):
/approve
/lgtm

The fallback approach LGTM.  This is consistent with how we handle "not found" in other core contrllers.

Thanks!

dims commented:
@jpbetz does this still look good? (lgtm got removed)

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:
```diff
@@ -381,6 +402,53 @@ func CollectParams(
+
+	mapping, err := restMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find REST mapping for %v: %w", gvk, err)
```
I'm concerned that if a CRD is deleted or not yet created this error might change how missing params are handled in a breaking way.

In the above call site of `getParamDirectly`there is a switch statement handling the various error cases.  The behavior before this change is for param to be `nil` if the param is not found. Are there error types that can be returned from `restMapper.RESTMapping` that should be treated as not found (`meta.IsNoMatchError`) to ensure we don't start returning errors in cases that would previously had resulted in a `nil` param?

Let's add a test for the case of a param that is a reference to a CR for a CRD that does not exist.

afipay replied:
@jpbetz Good point, I've addressed both
please help to review: https://github.com/kubernetes/kubernetes/pull/134423/commits/167b2fd132ed56b4249104446f6bf0ede26b5550

jpbetz commented:
> @jpbetz does this still look good? (lgtm got removed)

I think I'd missed an edge before. I've left a comment. LGTM after that is resolved.

jpbetz reviewed (approved):
/approve
/lgtm

pacoxu commented:
> Adding label `do-not-merge/contains-merge-commits` because PR contains merge commits, which are not allowed in this repository. Use `git rebase` to reapply your commits on top of the target branch. Detailed instructions for doing so can be found [here](https://git.k8s.io/community/contributors/guide/github-workflow.md#4-keep-your-branch-in-sync).
> 
> Details
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md). If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository.

This needs to be addressed. 

/release-note-none
(add a release note if needed)

afshin-paydar committed:
Fix param resolution race in ValidatingAdmissionPolicy
Add client fallback to handle race condition where param resources
exist in etcd but haven't propagated to informer cache yet. This
occurs when Policy, Binding, and param (e.g., ConfigMap) are created
in rapid succession during batch operations.

afipay commented:
made it a single commit, no changes after last LGTM

chrischdi commented:
Did also just hit. Is there something I can help with to get this over the line?

Maybe @jpbetz :-)

jpbetz commented:
cc @lalitc375

lalitc375 commented:
There is no diff after the last lgtm. 

/lgtm

jpbetz commented:
> can we get that picked to 1.36 too ?

Unfortunately, I don't think this meet the current cherrypick guidelines. The issues I see:

- The feature has been stable with this defect since the feature was introduced, so this is not a regression
- The patch is non-trivial, which introduces risk
